### ISSUE TITLE
Type System

### ISSUE THREAD

**Issue by NicholasCorrado** [UWHustle/hustle#28]
The operator code uses switch blocks in several places to properly handle different data types. We should have a type system that [replaces conditionals with polymorphism](https://refactoring.com/catalog/replaceConditionalWithPolymorphism.html) (also see [this repo](https://github.com/UWQuickstep/gift)) or we could use templates.

**Comment by pateljm** [comment]
Does arrow support: Int, double, string, date, timestamp, boolean? If yes, we can do 3.

**Comment by GindaChen** [comment]
@pateljm The answer is Yes! Arrow type system natively supports all these types.

@srsuryadev @johnchronis Let me know if you guys have suggestions and comments (particularly historical reasons why we did use that).

**Comment by johnchronis** [comment]
We did have the hustle type to be able to support custom data types (i.e. IP addrress was the running example at the time)
Essentially we wanted to be able to handle IP data, store them internally as ints but:
-despite being stored as ints when printed it should be formatted as IPS
-maybe be able to compare them with different semantics (or apply non-standard ops, like is_in_the_same_subnet(ip1, ip2)

Those operations would have been handled by the hustletype.
 
In short, the reason for having hustletypes would be to be able to extend to new  custom data types, if we do not want that functionality or can handle that in arrow somehow then we don't need to have hustle types

**Comment by pateljm** [comment]
> In short, the reason for having hustletypes would be to be able to extend to new custom data types, if we do not want that functionality or can handle that in arrow somehow then we don't need to have hustle types

Good memory @johnchronis -- I think we can scale back on the custom types and go to basic types in Arrow now. Get the basics right first and keep it simple.

**Comment by GindaChen** [comment]
> support custom data types (i.e. IP addrress was the running example at the time)
@johnchronis Im actually pretty confident that we can use the `arrow::Extention` type in arrow to implement IP Address. In fact, this mysterious `Extention` type is designed for user-defined type. The extention type can, in itself, defines its relationship with other types (ideally, of course, using what we have proposed ini #94). 

In fact, arrow had a test `IPv4` type in their code base:
https://github.com/apache/arrow/blob/2d140c3eeecca3ff7823edc8c9562ebd6a1c336a/cpp/src/arrow/array/array_view_test.cc#L55-L78

We can open up an issue to address user-defined type - will need to experiment a little bit on this.
